src/renderer/components/app.tsx

```diff
@@ -145,21 +147,27 @@ export class App extends React.Component {
     return routes;
   }
 
-  renderExtensionTabLayoutRoutes() {
+  renderExtensionTabLayoutRoutes(): JSX.Element[] {
     return clusterPageMenuRegistry.getRootItems().map((menu, index) => {
-      const tabRoutes = this.getTabLayoutRoutes(menu);
+      let route = this.extensionRoutes.get(menu);
+      if (route) {
+        return route;
+      }
 
+      const tabRoutes = this.getTabLayoutRoutes(menu);
       if (tabRoutes.length > 0) {
         const pageComponent = () => <TabLayout tabs={tabRoutes} />;
-
-        return <Route key={`extension-tab-layout-route-${index}`} component={pageComponent} path={tabRoutes.map((tab) => tab.routePath)} />;
+        route = <Route key={`extension-tab-layout-route-${index}`} component={pageComponent} path={tabRoutes.map((tab) => tab.routePath)} />;
+        this.extensionRoutes.set(menu, route);
```

Comment:
Probably this better to do not in `render()`. Also what if menu item tabs would change, but it's cached already?

Reply:
> Probably this better to do not in `render()`. 

What alternatives do we have?

> Also what if menu item tabs would change, but it's cached already?

We don't currently support dynamic menus (from an extension). If extension is updated then `menu` will be a different instance and it might cause issues (duplicate route?). Still I feel that it's bigger issue to recreate route on each re-render.

Reply:
@ixrock now it's refactored to use `reaction()` which happens outside of `render()`.